interp/iostream.go

```diff
@@ -0,0 +1,224 @@
+// Any other type of error returns -1 and err.
+//
+// The result mimicks gawk for expected child process errors:
+// 1. Returns the exit status of the child process and nil error on normal process exit.
+// 2. Returns 256 + signal on unhandled signal exit.
+// 3. Returns 512 + signal on unhandled signal exit which caused a core dump.
+func waitExitCode(cmd *exec.Cmd) (int, error) {
+	err := cmd.Wait()
+	if err == nil {
+		return 0, nil
+	}
+	ee, ok := err.(*exec.ExitError)
+	if !ok {
+		// Wait() returned an io error.
+		return -1, err
+	}
+	status, ok := ee.ProcessState.Sys().(syscall.WaitStatus)
+	if !ok {
+		// Maybe not all platforms support WaitStatus?
+		return -1, err
+	}
+	switch {
+	case status.CoreDump():
+		return 512 + int(status.Signal()), nil
+	case status.Signaled():
+		return 256 + int(status.Signal()), nil
+	case status.Exited():
+		return status.ExitStatus(), nil
+	default:
+		return -1, err
+	}
+}
+
+type inputStream interface {
+	io.ReadCloser
+	ExitCode() int
+}
+
+type outputStream interface {
+	io.WriteCloser
+	Flush() error
+	ExitCode() int
+}
+
+type outFileStream struct {
+	*bufio.Writer
+	closer   io.Closer
+	exitCode int
+	closed   bool
+}
+
+func newOutFileStream(wc io.WriteCloser, size int) outputStream {
+	b := bufio.NewWriterSize(wc, size)
+	return &outFileStream{b, wc, notClosedExitCode, false}
+}
+
+func (s *outFileStream) Close() error {
+	if s.closed {
+		return doubleCloseError
+	}
+	s.closed = true
+	flushErr := s.Writer.Flush()
+	closeErr := s.closer.Close()
+	if err := firstError(flushErr, closeErr); err != nil {
+		s.exitCode = -1
+		return err
+	}
+	s.exitCode = 0
+	return nil
+}
+
+func (s *outFileStream) ExitCode() int {
+	return s.exitCode
+}
+
+type outCmdStream struct {
+	*bufio.Writer
+	closer   io.Closer
+	cmd      *exec.Cmd
+	exitCode int
+	closed   bool
+}
+
+func newOutCmdStream(cmd *exec.Cmd) (outputStream, error) {
+	w, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, newError("error connecting to stdin pipe: %v", err)
+	}
+	err = cmd.Start()
+	if err != nil {
+		w.Close()
+		return nil, err
+	}
+	out := &outCmdStream{bufio.NewWriterSize(w, outputBufSize), w, cmd, notClosedExitCode, false}
+	return out, nil
+}
+
+func (s *outCmdStream) Close() error {
+	if s.closed {
+		return doubleCloseError
+	}
+	s.closed = true
+	flushErr := s.Writer.Flush()
+	closeErr := s.closer.Close()
+	var waitErr error
+	s.exitCode, waitErr = waitExitCode(s.cmd)
+	return firstError(waitErr, flushErr, closeErr)
+}
+
+func (s *outCmdStream) ExitCode() int {
+	return s.exitCode
+}
+
+// An outNullStream allows writes to not do anything while fulfilling the outputStream interface.
+type outNullStream struct {
+	io.Writer
+	closed bool
+}
+
+func newOutNullStream() outputStream { return &outNullStream{io.Discard, false} }
+func (s outNullStream) Flush() error { return nil }
+func (s *outNullStream) Close() error {
+	if s.closed {
+		return doubleCloseError
+	}
+	s.closed = true
+	return nil
+}
```

Comment:
I'm not sure this is necessary for `outNullStream` -- double-close it *not* an error for that type, right? But okay to leave it for consistency if you prefer.

Reply:
This is necessary to prevent the `outputStream` from being a leaky abstraction. The point is to produce the same behavior no matter which underlying implementation is used for the interface.